Rename internal `int.jl` type tuples and unions for clarity

eschnett opened the pull request:
`Base/int.jl` uses internally certain tuples and unions of various `[U]Int{8,116,32,64,128}`. I found the current naming scheme somewhat confusing; e.g. `IntTypes` was a tuple whereas  `SignedIntTypes` was a union.

Rename these for consistency:
- Invent name `BitInteger` for all these types
- For tuples, use suffix `Types` (used only for iterating over types)
- For unions, use neither `Union` prefix nor `Types` suffix
- Add some missing cases
- When using these, prefer using unions over splatting tuples.
- Clean up some places where these can be used naturally, e.g. in respective test cases.

eschnett commented:
Question: Would it make sense to omit the tuples, and to iterate over the unions' `types` field instead?

hayd commented:
Should these be deprecated? (Are they used in the wild?)

eschnett commented:
They are only used internally and are not exported.

eschnett commented:
Ping -- this is an internal cleanup only, no user-visible change.

StefanKarpinski commented:
I'm fine with these renames – the new names seem clearer. I do think that omitting the tuples and just doing everything with unions would be fine. When you need to iterate over a set of types in a union, you can just use `T.types`.  cc @JeffBezanson

tkelman commented:
How "internal" is the `.types` field considered to be? Hopefully these are all well covered enough that we'd notice if anything in the internal representation changed.

StefanKarpinski commented:
Since this change is strictly internal, I think it should be ok to use `.types` even if it's considered internal.

eschnett committed:
`Base/int.jl` uses internally certain tuples and unions of various [U]Int{8,116,32,64,128}. I found the current naming scheme somewhat confusing; e.g. IntTypes was a tuple whereas SignedIntTypes was a union.

Rename these for consistency:

- Invent name BitInteger for all these types
- For unions, use neither Union prefix nor Types suffix
- Replace tuples by `.types` field of union
- Add some missing cases

- When using these, prefer using unions over splatting tuples.
- Clean up some places where these can be used naturally, e.g. in respective test cases.

eschnett commented:
Removed tuples as discussed.

eschnett commented:
Rebased, ready to go.

vtjnash commented on base/int.jl:
```diff
@@ -204,9 +201,9 @@ rem{T<:Integer}(x::T, ::Type{T}) = x
     box(T,checked_fptosi(T,unbox(Tf,x)))
-convert{T<:UnionU64Types,Tf<:Union{Float32,Float64}}(::Type{T}, x::Tf) =
+convert{T<:BitUnsigned64,Tf<:Union{Float32,Float64}}(::Type{T}, x::Tf) =
     box(T,checked_fptoui(T,unbox(Tf,x)))
```
@JeffBezanson fyi: the rename on these lines will cause a rebase conflict with the correction to the function signature at: https://github.com/JuliaLang/julia/pull/13412/files#diff-d5b98c05f9bca9e586d65cfc16c83833R207

JeffBezanson commented:
I might revert this. A type Union is not a good way to implement a container of types, because (1) a union is not a list of types but a union of them, and therefore brings in type-level semantics that you might not want, and (2) I plan to change the representation and remove the `.types` field.

eschnett commented:
@JeffBezanson I will submit a new pull request that remedies this.

JeffBezanson commented:
Ok, since I'm fine with the renaming part of it.
